sourcing/domain/service.go

```diff
@@ -0,0 +1,318 @@
+	}
+
+	// AvailableSourcesProvider interface for DefaultSourcingService
+	AvailableSourcesProvider interface {
+		GetPossibleSources(ctx context.Context, product domain.BasicProduct, deliveryInfo *cartDomain.DeliveryInfo) ([]Source, error)
+	}
+
+	// StockProvider interface for DefaultSourcingService
+	StockProvider interface {
+		GetStock(ctx context.Context, product domain.BasicProduct, source Source) (int, error)
+	}
+)
+
+var (
+	_ SourcingService = new(DefaultSourcingService)
+
+	// ErrInsufficientSourceQty - use to indicate that the requested qty exceeds the available qty
+	ErrInsufficientSourceQty = errors.New("Available Source Qty insufficient")
+
+	// ErrNoSourceAvailable - use to indicate that no source for item is available at all
+	ErrNoSourceAvailable = errors.New("No Available Source Qty")
+
+	// ErrNeedMoreDetailsSourceCannotBeDetected - use to indicate that informations are missing to determine a source
+	ErrNeedMoreDetailsSourceCannotBeDetected = errors.New("Source cannot be detected")
+)
+
+//Inject the dependencies
+func (d *DefaultSourcingService) Inject(
+	logger flamingo.Logger,
+	dep *struct {
+		AvailableSourcesProvider AvailableSourcesProvider `inject:",optional"`
+		StockProvider            StockProvider            `inject:",optional"`
+	},
+) *DefaultSourcingService {
+	d.logger = logger.WithField(flamingo.LogKeyModule, "sourcing").WithField(flamingo.LogKeyCategory, "DefaultSourcingService")
+
+	if dep != nil {
+		d.availableSourcesProvider = dep.AvailableSourcesProvider
+		d.stockProvider = dep.StockProvider
+	}
+
+	return d
+}
+
+// GetAvailableSources - see description in Interface
+func (d *DefaultSourcingService) GetAvailableSources(ctx context.Context, product domain.BasicProduct, deliveryInfo *cartDomain.DeliveryInfo, decoratedCart *decorator.DecoratedCart) (AvailableSources, error) {
+	if err := d.checkConfiguration(); err != nil {
+		return nil, err
+	}
+
+	sources, err := d.availableSourcesProvider.GetPossibleSources(ctx, product, deliveryInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	var lastStockError error
+	availableSources := AvailableSources{}
+	for _, source := range sources {
+		qty, err := d.stockProvider.GetStock(ctx, product, source)
+		if err != nil {
+			d.logger.Error(err)
+			lastStockError = err
+
+			continue
+		}
+		if qty > 0 {
+			availableSources[source] = qty
+		}
+	}
+
+	// if a cart is given we need to deduct the possible allocated items in the cart
+	if decoratedCart != nil {
+		allocatedSources, err := d.AllocateItems(ctx, decoratedCart)
+		if err != nil {
+			return nil, err
+		}
+
+		itemIdsWithProduct := getItemIdsWithProduct(decoratedCart, product)
+
+		for _, itemID := range itemIdsWithProduct {
+			availableSources = availableSources.Reduce(allocatedSources[itemID].AllocatedQtys)
+		}
+	}
+
+	if len(availableSources) == 0 && lastStockError != nil {
+		return availableSources, errors.Wrap(ErrNoSourceAvailable, lastStockError.Error())
+	} else if len(availableSources) == 0 {
```

Comment:
no else necessary here because of the return statement

Reply:
Refactored to have it a bit better readable